internal/systemd/systemd_test.go

```diff
@@ -0,0 +1,170 @@
+package systemd_test
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"testing"
+
+	"github.com/godbus/dbus/v5"
+	"github.com/godbus/dbus/v5/introspect"
+	log "github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/require"
+	"github.com/ubuntu/adsys/internal/consts"
+	"github.com/ubuntu/adsys/internal/systemd"
+	"github.com/ubuntu/adsys/internal/testutils"
+)
+
+var ctx = context.Background()
+
+func TestManageUnit(t *testing.T) {
+	t.Parallel()
+
+	bus := testutils.NewDbusConn(t)
+
+	systemdCaller, err := systemd.New(bus)
+	require.NoError(t, err, "Setup: failed to create systemd caller")
+
+	tests := map[string]struct {
+		unitName string
+		action   string
+
+		wantErr bool
+	}{
+		"Start unit that exists":   {action: "start"},
+		"Stop unit that exists":    {action: "stop"},
+		"Enable unit that exists":  {action: "enable"},
+		"Disable unit that exists": {action: "disable"},
+
+		// Error cases
+		"Error when starting unit that doesn't exist": {unitName: absentUnit, action: "start", wantErr: true},
+		"Error when starting failing unit":            {unitName: failingUnit, action: "start", wantErr: true},
+
+		"Error when stopping unit that doesn't exist": {unitName: absentUnit, action: "stop", wantErr: true},
+		"Error when stopping failing unit":            {unitName: failingUnit, action: "stop", wantErr: true},
+
+		"Error when enabling unit that doesn't exist":  {unitName: absentUnit, action: "enable", wantErr: true},
+		"Error when disabling unit that doesn't exist": {unitName: absentUnit, action: "disable", wantErr: true},
+	}
+
+	for name, tc := range tests {
+		tc := tc
+		name := name
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+
+			if tc.unitName == "" {
+				tc.unitName = "existing-service.service"
+			}
+
+			var err error
+			switch tc.action {
+			case "start":
+				err = systemdCaller.StartUnit(ctx, tc.unitName)
+			case "stop":
+				err = systemdCaller.StopUnit(ctx, tc.unitName)
+			case "enable":
+				err = systemdCaller.EnableUnit(ctx, tc.unitName)
+			case "disable":
+				err = systemdCaller.DisableUnit(ctx, tc.unitName)
+			default:
+				panic("unknown systemd action")
+			}
+
+			if tc.wantErr {
+				require.Error(t, err, fmt.Sprintf("Action %s should have failed but it didn't", tc.action))
+				return
+			}
+			require.NoError(t, err, fmt.Sprintf("Action %s shouldn't have failed but it did", tc.action))
+		})
+	}
+}
+
+func TestDaemonReload(t *testing.T) {
+	t.Parallel()
+
+	bus := testutils.NewDbusConn(t)
+
+	systemdCaller, err := systemd.New(bus)
+	require.NoError(t, err, "Setup: failed to create systemd caller")
+
+	err = systemdCaller.DaemonReload(ctx)
+	require.NoError(t, err, "DaemonReload shouldn't have failed but it did")
+
+	s.setErrorOnReload()
```

Comment:
That’s true, BUT, if here, you are mixing the 2 tests in a potential single error (in addition to make it less extensible). Do you think this will be too much boilerplate to do like the other one?

Reply:
I have a strong feeling, but I can’t really argue enough with the assert (apart that you need to look at the test if the first fail alone or not :p), so let’s go with this (even if meh :p).